source/common/config/grpc_mux_impl.cc

```diff
@@ -250,15 +251,22 @@ void GrpcMuxImpl::updateControlPlaneConfig(const envoy::api::v2::DiscoveryRespon
         auto* new_config_source_control_info =
             (*control_plane_config->mutable_service_control_plane_info())[xds_service_]
                 .add_config_source_control_plane();
-        populateControlPlaneConfigSource(control_plane_config.get(), new_config_source_control_info,
+        populateControlPlaneConfigSource(new_config_source_control_info,
                                          message.control_plane().identifier());
       } else {
         // update the control plane identifier of existing grpc_service.
+        uint32_t config_source_index = -1;
+        for (auto& config_control_plane :
+             service_control_plane_info->second.config_source_control_plane()) {
+          config_source_index++;
+          if (message_differencer.Compare(grpc_service_, config_control_plane.grpc_service())) {
+            break;
+          }
+        }
         auto* existing_config_source_control_info =
             (*control_plane_config->mutable_service_control_plane_info())[xds_service_]
-                .mutable_config_source_control_plane(config_source_index_);
-        populateControlPlaneConfigSource(control_plane_config.get(),
-                                         existing_config_source_control_info,
+                .mutable_config_source_control_plane(config_source_index);
+        populateControlPlaneConfigSource(existing_config_source_control_info,
```

Comment:
Thinking about how this can be simplified. What if we avoid having `GrpcMuxImpl` be in the business of doing low level technical protobuf mangling, and instead we have it supply `ConfigTracker` a lambda and whatever identifier is needed, and `ConfigTracker` can call this lambda when it needs updated state?

Reply:
`ConfigTracker` may have to keep track of all the lambdas (for each of the grpc service + combinations) and for that it has to maintain a parallel structure and call those lambdas at appropriate places. It might simplify a bit but not that much I guess.

Reply:
This seems pretty reasonable division of labor though, since`GrpcMuxImpl` is in the business of muxing xDS over gRPC and. ConfigTracker is in the business of tracking configuration related things. My main concern today is that we have a lot of low level proto surgery happening in `GrpcMuxImpl`.